validator/validator-main.protoascii

```diff
@@ -4638,6 +4674,42 @@ tags: {
   }
 }
 
+tags: {
+  html_format: AMP
+  enabled_by: "transformed"
+  tag_name: "IMG"
+  spec_name: "amp-img > img (transformed)"
+  # Ideally we'd be able to use regular amp-img parent, but runtime currently
+  # requires an i-amphtml-ssr attribute on the amp-img. We could do away with
+  # that, though.
+  mandatory_parent: "amp-img (transformed)"
+  attrs: { name: "alt" }
+  attrs: { name: "attribution" }
+  attrs: { name: "object-fit" }
+  attrs: { name: "object-position" }
+  attrs: { name: "referrerpolicy" }
+  attrs: { name: "sizes" }
+  attrs: { name: "title" }
+  attr_lists: "mandatory-src-or-srcset"
+  # SSR requires these explicit attributes
+  attrs: {
+    name: "class"
+    value_regex: "i-amphtml-fill-content\\s+i-amphtml-replaced-content|"
+        "i-amphtml-replaced-content\\s+i-amphtml-fill-content"
+    mandatory: true
+  }
+  attrs: {
+    name: "decoding"
+    value: "async"
+    mandatory: true
+  }
+  attrs: {
+    name: "loading"
+    value: "lazy"
```

Comment:
Did we test whether loading=lazy still triggers preload scanner to fetch the hero image pre-layout?

Reply:
Oh, this is a good point. Testing with https://output.jsbin.com/husuxij/quiet:

| Browser            | Loading | Pre-scan? |
|--------------------|---------|-----------|
| Chrome 83          | Lazy    | ❌         |
| ^                  | Eager   | ✅         |
| Safari TP 109      | Lazy    | ❌         |
| ^                  | Eager   | ✅         |
| Firefox 80 Nightly | Lazy    | ❌         |
| ^                  | Eager   | ❌         |

@sebastianbenz: I'm assuming this means we need to use non-lazy loading for SSR images?

Reply:
We should probably only allow 1-2 non-lazy images per doc. I'm not sure how much sense it makes to add the lazy images at all. I think the simplest would be to make these non-lazy and only allow a small amount.

Reply:
>  preload scanner to fetch the hero image pre-layout

for my own edification: Is this relevant in the case of SSR without boilerplate?

Reply:
Yep, the impact is actually roughly the same in both cases as the boilerplate does not block layout (it just blocks visibility of content). You still get a win without the prescanner, as you can get downloads before JS loads, however, you'd expect a big win with the prescanner because

- layout can be expensive
- AMP's JS may run before layout when it is cached

Reply:
Thanks. I've updated the [AMP Optimizer PR](https://github.com/ampproject/amp-toolbox/pull/859#pullrequestreview-448368134) to use `loading=eager` and only optimize a maximum of two images.

Reply:
I'd advocate for keeping it unspecified for now

Reply:
@twifkak brought up using `<link rel=preload>` on these lazy images, and that appears to fix the prescanner for us (https://output.jsbin.com/nerokig/4/quiet):

| Browser            | Loading | Pre-scan? |
|--------------------|---------|-----------|
| Chrome 83          | Lazy link rel    | ✅         |
| ^                  | Eager   | ✅         |
| Safari TP 109      | Lazy link rel    | ✅         |
| ^                  | Eager   | ✅         |
| Firefox 80 Nightly | Lazy link rel    | ✅         |
| ^                  | Eager   | ❌         |

Reply:
Outside of Chrome, this would only work for images without srcset due to lack of support for `imagesrcset`.

Reply:
There's currently a `imagesrcset` transformer that rewrite them to multiple `<link rel=preload media="…">`  with mutually exclusive media queries. We need to expand this to support `imagesizes`, but I think we can use the same general idea for better support.

Reply:
I've given feedback before that while that might seem to work it is a very dangerous game. E.g. a browser in data-saver mode might choose to pick a smaller srcset value than the media query would yield leading to double download.

Why can't we generate `<img srcset="…" sizes="…" style="display:none">` in place of `<link rel=preload as=image>`?

Reply:
Re: @jridgewell's [comment from Jul 14](https://github.com/ampproject/amphtml/pull/29025#discussion_r454724219), if I set the script to async and set body visibility to hidden, the lazy image is fetched immediately. See https://output.jsbin.com/nowisumiwu/quiet.

Are the pre-`visibility: visible` AMP layouts a useful approximation of what images will be above the fold post full layout? If so, there could be value in setting `<img loading=lazy>` on all images. Then we let the browser decide which images to fetch/decode before AMP layout has finished.

Could be combined with hero image detection for a more explicit `<link rel=preload>` (or `<img style="display:none">` per @cramforce's last comment) where we're confident.

On cache side, it would need to be removed for privacy, but still useful on-origin and on-SXG.

Reply:
correction: Wouldn't need to be removed for privacy, since they'd be rewritten URLs. Might want to remove to reduce data usage, for when navigation doesn't happen.